scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx

```diff
@@ -0,0 +1,294 @@
+
+    constructor(props: IProps) {
+        super(props);
+        this.state = {
+            currentTime: 0,
+            thumbnailsRender: [],
+            trim: {
+                start: this.props.trim.start,
+                end: this.props.trim.end,
+            },
+        };
+        this.controlbar = React.createRef();
+        this.intervalTimer = 0;
+        this.positionX = 0;
+        this.handleDrag = this.handleDrag.bind(this);
+        this.handleTimelineClick = this.handleTimelineClick.bind(this);
+        this.handleDragOver = this.handleDragOver.bind(this);
+        this.handleDragEnd = this.handleDragEnd.bind(this);
+        this.tick = this.tick.bind(this);
+
+        this.handleTimelineClick = this.handleTimelineClick.bind(this);
+    }
+
+    componentDidMount() {
+        // call tick every 100ms to update current time state
+        // Don't use event timeupdate, because cannot set timeloop, default is 250ms
+        this.intervalTimer = window.setInterval(this.tick, 100);
+        // Use dragover event window to save position X, because event.clientX can not get in Firefox
+        document.addEventListener('dragover', this.handleDragOver);
+        this.setThumbnailsRender();
+    }
+
+    componentDidUpdate(prevProps: IProps) {
+        if (prevProps.thumbnails !== this.props.thumbnails) {
+            this.setThumbnailsRender();
+        }
+        if (this.props.trim !== prevProps.trim) {
+            this.updateTrim(this.props.trim.start, this.props.trim.end);
+        }
+    }
+
+    componentWillUnmount() {
+        clearInterval(this.intervalTimer);
+        document.removeEventListener('dragover', this.handleDragOver);
+    }
+
+    updateTrim(start: number, end: number) {
+        this.setState({
+            trim: {
+                start: start,
+                end: end,
+            },
+        });
+    }
+
+    handleDrag(type: 'left' | 'right') {
+        let time = this.getPositionInBar(this.positionX) * this.props.video.duration;
+
+        if (type === 'left' && time < this.state.trim.end - VIDEO_MIN_DURATION) {
+            this.updateTrim(time, this.state.trim.end);
+        }
+        if (type === 'right' && time > this.state.trim.start + VIDEO_MIN_DURATION) {
+            this.updateTrim(this.state.trim.start, time);
+        }
+    }
+
+    handleDragOver(e: DragEvent | React.DragEvent<HTMLDivElement>) {
+        if (e.clientX) {
+            this.positionX = e.clientX;
+        }
+    }
+
+    // select a number of thumbnails will be rendered based on screen width
+    // because there is not enough space for all of them
+    setThumbnailsRender() {
+        if (this.props.thumbnails.length === 0) {
+            this.setState({thumbnailsRender: []});
+            return;
+        }
+
+        const width = this.props.thumbnails[0].width;
+        // Plus one more incomplete frame because we do not know thumbnails width beforehand to
+        // make timeline render with just enough width space
```

Comment:
Video aspect ratio & browser screen size are known so thumbnail size could be calculated, no? Even if not, why not delay rendering of the component until the thumbnails are downloaded and sizes are known?